Wrapper startup broken on macOS

As first discussed in https://github.com/eXist-db/exist/pull/1065#issuecomment-232386003, something about the changes in https://github.com/eXist-db/exist/pull/1016 (merged in on May 18) resulted in wrapper startup failures on macOS with builds of the develop branch since that day.  As a result, methods of startup which used to work - double-clicking on the macOS `.app` version of eXist (from a .dmg installer), double-clicking on `start.jar`, or using the command `open start.jar` - no longer work.

To confirm this, I downloaded two nightlies from Adam's site to see whether this change resulted in the problem we're seeing. Indeed, the .app from the .dmg nightly of May 17 starts right up, whereas the May 19th version doesn't start.

I also see a difference between the process of the May 17th (158eaa3) and May 19th (50d037f) versions. Here's how each appears when I start them via start.jar (opening its application package and double-click directly on its start.jar):

**May 17:**

> $ ps aux | grep java
> joe              34956   0.2  2.8  7290824 467980   ??  S     5:06PM   0:18.43 /Library/Internet Plug-Ins/JavaAppletPlugin.plugin/Contents/Home/bin/java -Dwrapper.java.library.path=/Users/joe/Desktop/eXist-db.app/Contents/Resources/eXist-db/.tools/wrapper/ -Xms256m -Dfile.encoding=UTF-8 -Xmx1024m -Xdock:name=eXist-db -Xdock:icon=icon.png -Dapple.laf.useScreenMenuBar=true -Dapple.awt.UIElement=true -Dexist.home=/Users/joe/Desktop/eXist-db.app/Contents/Resources/eXist-db/. -Djava.endorsed.dirs=/Users/joe/Desktop/eXist-db.app/Contents/Resources/eXist-db/./lib/endorsed -jar /Users/joe/Desktop/eXist-db.app/Contents/Resources/eXist-db/./start.jar org.exist.launcher.Launcher

**May 19:**

> $ ps aux | grep java
> joe              34935   0.0  0.1  8284660  17660   ??  Ss    5:03PM   0:00.02 /Library/Internet Plug-Ins/JavaAppletPlugin.plugin/Contents/Home/bin/java -jar /Users/joe/Desktop/eXist-db.app/Contents/Resources/eXist-db/start.jar

When I swap the May 17th version out and put the May 19th version in its place, the startup command from the May 17th process output successfully starts the May 19th version:

> $ /Library/Internet\ Plug-Ins/JavaAppletPlugin.plugin/Contents/Home/bin/java -Dwrapper.java.library.path=/Users/joe/Desktop/eXist-db.app/Contents/Resources/eXist-db/.tools/wrapper/ -Xms256m -Dfile.encoding=UTF-8 -Xmx1024m -Xdock:name=eXist-db -Xdock:icon=icon.png -Dapple.laf.useScreenMenuBar=true -Dapple.awt.UIElement=true -Dexist.home=/Users/joe/Desktop/eXist-db.app/Contents/Resources/eXist-db/. -Djava.endorsed.dirs=/Users/joe/Desktop/eXist-db.app/Contents/Resources/eXist-db/./lib/endorsed -jar /Users/joe/Desktop/eXist-db.app/Contents/Resources/eXist-db/./start.jar org.exist.launcher.Launcher

> Are specific to your Mac because of the file type associations configured on your machine.

@adamretter Hmm, if so, then I'm confused as to why, with 3.0RC1 (dmg release), does double-clicking its start.jar does start the splash screen and task bar icon.

> Regards building your own `.app`, I believe that the process would be `./build.sh clean && ./build.sh && ./build.sh dist && ./build.sh app`

Ah, I hadn't performed the `dist` step. And, in fact, running this command, the build fails on the `dist` step:

> BUILD FAILED
> /Users/joe/workspace/exist/build/scripts/dist.xml:331: /Users/joe/workspace/exist/dist/eXist-db-3.0.RC2/yajsw/bin does not exist.

--- a/build/scripts/macosx.xml
+++ b/build/scripts/macosx.xml
@@ -5,13 +5,13 @@
 <!-- ======================================================================= -->
 <!-- $Id$ -->
 <project basedir="../.." default="all" name="Mac OS App">
-    
+
     <description>Build installer</description>
-    
+
     <!-- import common targets -->
     <import file="../../build.xml"/>
     <import file="git-support.xml"/>
-    
+
     <property name="dist" value="${basedir}/dist"/>
     <property name="app.dir" value="${dist}/eXist-db.app"/>
     <property name="app.resources" value="${app.dir}/Contents/Resources"/>
@@ -22,26 +22,26 @@
     <property name="appbundler.url" value="http://java.net/downloads/appbundler/appbundler-1.0.jar"/>
 
     <available property="appbundler.available" file="${appbundler.jar}"/>
-    
-    <target name="app" description="Build Mac OS X dmg (unsigned)" depends="bundle,copy-all,dmg"/>
-    
-    <target name="app-signed" description="Build Mac OS X dmg (signed)" depends="bundle,copy-all,codesign,dmg"/>
-    
+
+    <target name="app" description="Build Mac OS X dmg (unsigned)" depends="all,bundle,copy-all,dmg"/>
+
+    <target name="app-signed" description="Build Mac OS X dmg (signed)" depends="all,bundle,copy-all,codesign,dmg"/>
+
     <target name="install-appbundler" unless="appbundler.available">
         <taskdef name="fetch" classname="nl.ow.dilemma.ant.fetch.FetchTask">
             <classpath>
                 <pathelement location="${asocat-exist.jar}"/>
             </classpath>
         </taskdef>
-        
+
         <fetch dest="${tools.ant}/lib" url="${appbundler.url}"
             failonerror="false" maxtime="120"/>
     </target>
-    
+
     <target name="bundle" depends="prepare-app">
-        <taskdef 
-            name="bundleapp" 
-            classname="com.oracle.appbundler.AppBundlerTask" 
+        <taskdef
+            name="bundleapp"
+            classname="com.oracle.appbundler.AppBundlerTask"
             classpath="${appbundler.jar}" />
         <bundleapp
             outputdirectory="${dist}"
@@ -51,7 +51,9 @@
             mainclassname="org.exist.start.Main"
             shortversion="${project.version.numeric}"
             icon="${app.icon}">
-            <classpath file="start.jar"/>
+            <classpath dir=".">
+                <include name="start.jar"/>
+            </classpath>
             <option value="-Dexist.home=$APP_ROOT/Contents/Resources/eXist-db"/>
         </bundleapp>
     </target>
@@ -198,7 +200,7 @@
 
         <!-- Set this property value to your application name -->
         <property name="app.name" value="${project.name}"/>
-        
+
         <property name="dmg.name" value="${app.name}-${project.version}-${git.commit}.dmg"/>
         <echo message="Creating ${dmg.name}..."/>
 
@@ -208,7 +210,7 @@
         <mkdir dir="${mountdir}"/>
 
         <!-- Delete previously created DMG -->
-        <delete file="${dist}/${app.name}.dmg" quiet="yes" failonerror="false"/>
+        <delete file="${dist}/${dmg.name}" quiet="yes" failonerror="false"/>
 
         <!-- Create a temporary Disk Image (Mac) -->
         <exec executable="/usr/bin/hdiutil" os="Mac OS X" failonerror="true">
@@ -222,7 +224,7 @@
             <arg value="-format"/>
             <arg value="UDRW"/>
         </exec>
-        
+
         <!-- Attach the temporary image (Mac) -->
         <exec executable="/usr/bin/hdiutil" os="Mac OS X" failonerror="true">
             <arg value="attach"/>
@@ -230,7 +232,7 @@
             <arg value="-mountroot"/>
             <arg value="${mountdir}/"/>
         </exec>
-        
+
         <!-- Create a temporary Disk Image (Linux) -->
         <exec executable="du" outputproperty="du.out" os="Linux" failonerror="true">
             <arg value="-sm"/>
@@ -247,14 +249,14 @@
                 <math result="dmg.size" operand1="${du.size}" operation="add" operand2="2" datatype="int"/>
             </then>
         </if>
-        
+
         <exec executable="dd" os="Linux" failonerror="true">
             <arg value="if=/dev/zero"/>
             <arg value="of=${dist}/${app.name}-tmp.dmg"/>
             <arg value="bs=1M"/>
             <arg value="count=${dmg.size}"/>
         </exec>
-        
+
         <exec executable="/sbin/mkfs" os="Linux" failonerror="true">
             <arg value="-t"/>
             <arg value="hfsplus"/>
@@ -262,7 +264,7 @@
             <arg value="${app.name}"/>
             <arg value="${dist}/${app.name}-tmp.dmg"/>
         </exec>
-        
+
         <!-- Attach the temporary image (Linux) -->
         <exec executable="whoami" os="Linux" outputproperty="username" failonerror="true"/>
         <exec executable="sudo" os="Linux" failonerror="true">
@@ -278,11 +280,11 @@
             <arg value="${app.dir}"/>
             <arg value="${mountdir}"/>
         </exec>
-        
+
         <condition property="app.dest" value="${mountdir}/${app.name}" else="${mountdir}">
             <os family="mac"/>
         </condition>
-        
+
         <!-- Copy the background, the volume icon and DS_Store files -->
         <mkdir dir="${app.dest}/.DropDMGBackground"/>
         <copy file="installer/background.png"
@@ -310,7 +312,7 @@
             <arg value="detach"/>
             <arg value="${app.dest}"/>
         </exec>
-        
+
         <!-- Detach the temporary image (Linux) -->
         <exec executable="sudo" os="Linux" failonerror="true">
             <arg value="umount"/>
@@ -332,7 +334,7 @@
             <arg value="-yes"/>
             <arg value="${dist}/${dmg.name}"/>
         </exec>
-        
+
         <!-- Copy it to a new image (Linux) -->
         <exec executable="cp" os="Linux" failonerror="true">
             <arg value="${dist}/${app.name}-tmp.dmg"/>
@@ -342,11 +344,11 @@
         <!-- Delete the temporary image -->
         <delete file="${dist}/${app.name}-tmp.dmg"
                 quiet="yes" failonerror="false"/>
-        
+
         <!-- Delete the mount point -->
         <delete includeemptydirs="true">
             <fileset dir="${mountdir}" includes="**/*"/>
         </delete>
-        
+
     </target>
 </project>

--- a/installer/scripts/setup.bat
+++ b/installer/scripts/setup.bat
@@ -18,6 +18,9 @@ set JAVA_CMD="%JAVA_HOME%\bin\java"
 set JAVA_ENDORSED_DIRS="%EXIST_HOME%"\lib\endorsed
 set JAVA_OPTS="-Xms64m -Xmx768m -Djava.endorsed.dirs=%JAVA_ENDORSED_DIRS%"
 
+rem make sure there's the jetty tmp directory
+mkdir "%EXIST_HOME%\tools\jetty\tmp"
+
 rem echo "JAVA_HOME: %JAVA_HOME%"
 rem echo "EXIST_HOME: %EXIST_HOME%"
 echo %JAVA_OPTS%

--- a/src/org/exist/launcher/ConfigurationDialog.java
+++ b/src/org/exist/launcher/ConfigurationDialog.java
@@ -8,6 +8,7 @@ import java.nio.file.Paths;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Properties;
+import java.util.function.Consumer;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 import java.util.stream.Collectors;
@@ -35,7 +36,7 @@ import org.exist.util.FileUtils;
 
 public class ConfigurationDialog extends JDialog {
 
-    private final Launcher launcher;
+    private final Consumer<Boolean> callback;
     private boolean changed = false;
     private boolean dataDirChanged = false;
     private boolean jettyConfigChanged = false;
@@ -44,13 +45,13 @@ public class ConfigurationDialog extends JDialog {
     /**
      * Creates new form ConfigurationDialog
      */
-    public ConfigurationDialog(Launcher launcher) {
+    public ConfigurationDialog(Consumer<Boolean> callback) {
         setModal(true);
         setTitle("eXist-db System Configuration");
 
         initComponents();
 
-        this.launcher = launcher;
+        this.callback = callback;
         
         final PropertiesConfiguration vmProperties = LauncherWrapper.getVMProperties();
         final int maxMemProp = vmProperties.getInt("memory.max", 1024);
@@ -73,8 +74,8 @@ public class ConfigurationDialog extends JDialog {
             if (ports.containsKey("jetty.port")) {
                 httpPort.setValue(ports.get("jetty.port"));
             }
-            if (ports.containsKey("jetty.port.ssl")) {
-                sslPort.setValue(ports.get("jetty.port.ssl"));
+            if (ports.containsKey("jetty.ssl.port")) {
+                sslPort.setValue(ports.get("jetty.ssl.port"));
             }
         } catch (DatabaseConfigurationException ex) {
             Logger.getLogger(ConfigurationDialog.class.getName()).log(Level.SEVERE, null, ex);
@@ -516,13 +517,14 @@ public class ConfigurationDialog extends JDialog {
                 properties.clear();
                 properties.setProperty("port", httpPort.getValue().toString());
                 properties.setProperty("port.ssl", sslPort.getValue().toString());
-                ConfigurationUtility.saveConfiguration("tools/jetty/etc/jetty.xml", "jetty.xsl", properties);
+                ConfigurationUtility.saveConfiguration("tools/jetty/etc/jetty-ssl.xml", "jetty.xsl", properties);
+                ConfigurationUtility.saveConfiguration("tools/jetty/etc/jetty-http.xml", "jetty.xsl", properties);
             }
             if (beforeStart) {
                 beforeStart = false;
                 btnCancel.setVisible(true);
                 setVisible(false);
-                this.launcher.shutdown(true);
+                callback.accept(true);
             } else if (changed || dataDirChanged || jettyConfigChanged) {
                 int r = JOptionPane.showConfirmDialog(this, "Database needs to be restarted to apply the " +
                             "new settings.", "Confirm restart", JOptionPane.YES_NO_OPTION, JOptionPane.WARNING_MESSAGE);
@@ -530,7 +532,7 @@ public class ConfigurationDialog extends JDialog {
                     changed = false;
                     dataDirChanged = false;
                     setVisible(false);
-                    this.launcher.shutdown(true);
+                    callback.accept(true);
                 }
             }
         } catch (IOException | ConfigurationException e) {

--- a/src/org/exist/launcher/ConfigurationUtility.java
+++ b/src/org/exist/launcher/ConfigurationUtility.java
@@ -35,7 +35,14 @@ public class ConfigurationUtility {
 
     public static Map<String, Integer> getJettyPorts() throws DatabaseConfigurationException {
         final Map<String, Integer> ports = new HashMap<>();
-        final Path jettyConfig = ConfigurationHelper.lookup("tools/jetty/etc/jetty.xml");
+        final Path jettyHttpConfig = ConfigurationHelper.lookup("tools/jetty/etc/jetty-http.xml");
+        final Path jettyHttpsConfig = ConfigurationHelper.lookup("tools/jetty/etc/jetty-ssl.xml");
+        getJettyPorts(ports, jettyHttpConfig);
+        getJettyPorts(ports, jettyHttpsConfig);
+        return ports;
+    }
+
+    private static void getJettyPorts(Map<String, Integer> ports, Path jettyConfig) throws DatabaseConfigurationException {
         if (Files.exists(jettyConfig)) {
             try {
                 final XMLStreamReader reader = XMLInputFactory.newInstance().createXMLStreamReader(Files
@@ -44,7 +51,7 @@ public class ConfigurationUtility {
                     final int status = reader.next();
                     if (status == XMLStreamReader.START_ELEMENT && "SystemProperty".equals(reader.getLocalName())) {
                         final String name = reader.getAttributeValue(null, "name");
-                        if (name != null && name.startsWith("jetty.port")) {
+                        if (name != null && (name.equals("jetty.port") || name.equals("jetty.ssl.port"))) {
                             final String defaultValue = reader.getAttributeValue(null, "default");
                             if (defaultValue != null) {
                                 try {
@@ -60,7 +67,6 @@ public class ConfigurationUtility {
                 throw new DatabaseConfigurationException(e.getMessage(), e);
             }
         }
-        return ports;
     }
 
     public static void saveProperties(Properties properties) throws ConfigurationException, IOException {

--- a/src/org/exist/launcher/Launcher.java
+++ b/src/org/exist/launcher/Launcher.java
@@ -137,7 +137,7 @@ public class Launcher extends Observable implements Observer {
             initSystemTray = initSystemTray();
         }
 
-        configDialog = new ConfigurationDialog(Launcher.this);
+        configDialog = new ConfigurationDialog(this::shutdown);
 
         splash = new SplashScreen(this);
         splash.addWindowListener(new WindowAdapter() {

--- a/src/org/exist/launcher/LauncherWrapper.java
+++ b/src/org/exist/launcher/LauncherWrapper.java
@@ -37,7 +37,10 @@ import java.io.InputStreamReader;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Iterator;
+import java.util.List;
 
 /**
  * A wrapper to start a Java process using start.jar with correct VM settings.
@@ -54,7 +57,16 @@ public class LauncherWrapper {
 
     public final static void main(String[] args) {
         final LauncherWrapper wrapper = new LauncherWrapper(LAUNCHER);
-        wrapper.launch();
+        if (ConfigurationUtility.isFirstStart()) {
+            System.out.println("First launch: opening configuration dialog");
+            ConfigurationDialog configDialog = new ConfigurationDialog(restart -> {
+                wrapper.launch();
+            });
+            configDialog.open(true);
+            configDialog.requestFocus();
+        } else {
+            wrapper.launch();
+        }
     }
 
     protected String command;
@@ -71,20 +83,23 @@ public class LauncherWrapper {
         final OperatingSystem os = OperatingSystem.instance();
         final ProcessManager pm = os.processManagerInstance();
         final Process process = pm.createProcess();
-        final StringBuilder cmdLine = new StringBuilder()
-                .append(getJavaCmd())
-                .append(getJavaOpts(home, vmProperties));
+        final List<String> args = new ArrayList<>();
+        args.add(getJavaCmd());
+        getJavaOpts(args, home, vmProperties);
+
         if(Boolean.parseBoolean(debugLauncher)) {
-            cmdLine.append(" -Xdebug -Xrunjdwp:transport=dt_socket,server=y,suspend=y,address=4000");
+            args.add("-Xdebug");
+            args.add("-Xrunjdwp:transport=dt_socket,server=y,suspend=y,address=4000");
         }
-        cmdLine.append(" -jar start.jar ");
-        cmdLine.append(command);
+        args.add("-jar");
+        args.add("start.jar");
+        args.add(command);
+
         process.setWorkingDir(home);
         process.setVisible(false);
         process.setPipeStreams(false, false);
-        final String cmdLineStr = cmdLine.toString();
-        process.setCommand(cmdLineStr);
-        System.out.println(cmdLineStr);
+        System.out.println(String.join(" ", args));
+        process.setCommand(args.toArray(new String[args.size()]));
 
         if (process instanceof WindowsXPProcess) {
             ((WindowsXPProcess)process).startElevated();
@@ -113,43 +128,36 @@ public class LauncherWrapper {
         return "java";
     }
 
-    protected String getJavaOpts(String home, PropertiesConfiguration vmProperties) {
-        final StringBuilder opts = new StringBuilder();
-
-        opts.append(getVMOpts(vmProperties));
+    protected void getJavaOpts(List<String> args, String home, PropertiesConfiguration vmProperties) {
+        getVMOpts(args, vmProperties);
 
         if (command.equals(LAUNCHER) && "mac os x".equals(OS)) {
-            opts.append(" -Dapple.awt.UIElement=true");
+            args.add("-Dapple.awt.UIElement=true");
         }
-        opts.append(" -Dexist.home=");
-        opts.append('"').append(home).append('"');
+        args.add("-Dexist.home=\"" + home + '"');
 
-        opts.append(" -Djava.endorsed.dirs=");
-        opts.append('"').append(home).append("/lib/endorsed").append('"');
-
-        return opts.toString();
+        args.add("-Djava.endorsed.dirs=\"" + home + "/lib/endorsed\"");
     }
 
-    protected String getVMOpts(PropertiesConfiguration vmProperties) {
-        final StringBuilder opts = new StringBuilder();
+    protected void getVMOpts(List<String> args, PropertiesConfiguration vmProperties) {
         for (final Iterator<String> i = vmProperties.getKeys(); i.hasNext(); ) {
             final String key = i.next();
             if (key.startsWith("memory.")) {
                 if ("memory.max".equals(key)) {
-                    opts.append(" -Xmx").append(vmProperties.getString(key)).append('m');
+                    args.add("-Xmx" + vmProperties.getString(key) + 'm');
                 } else if ("memory.min".equals(key)) {
-                    opts.append(" -Xms").append(vmProperties.getString(key)).append('m');
+                    args.add("-Xms" + vmProperties.getString(key) + 'm');
                 }
             } else if ("vmoptions".equals(key)) {
-                opts.append(' ').append(vmProperties.getString(key));
+                args.add(vmProperties.getString(key));
             } else if (key.startsWith("vmoptions.")) {
                 final String os = key.substring("vmoptions.".length()).toLowerCase();
                 if (OS.contains(os)) {
-                    opts.append(' ').append(vmProperties.getString(key));
+                    final String value = vmProperties.getString(key);
+                    Arrays.stream(value.split("\\s+")).forEach(args::add);
                 }
             }
         }
-        return opts.toString();
     }
 
     public static PropertiesConfiguration getVMProperties() {

--- a/src/org/exist/launcher/jetty.xsl
+++ b/src/org/exist/launcher/jetty.xsl
@@ -10,7 +10,7 @@
     
     <xsl:template match="SystemProperty[@name='jetty.port']"><SystemProperty name="jetty.port" default="{$port}"/></xsl:template>
 
-    <xsl:template match="SystemProperty[@name='jetty.port.ssl']"><SystemProperty name="jetty.port.ssl" default="{$port.ssl}"/></xsl:template>
+    <xsl:template match="SystemProperty[@name='jetty.ssl.port']"><SystemProperty name="jetty.ssl.port" default="{$port.ssl}"/></xsl:template>
     
     <xsl:template match="@*|node()" priority="-1">
         <xsl:copy>

